### PR TITLE
Attempt to dial all addresses in parallel and use the first success

### PR DESCRIPTION
**Pull request by jackkleeman**
https://github.com/ipfs/js-ipfs/issues/812
We were not dialling all addresses in sequence, currently only the first address was ever dialled leading to regular failures. This dials all in parallel and uses the first success (or the first timeout).

### CODE REVIEW

File: src/transport.js
Comment (dignifiedquire) [inline src/transport.js:57]: I think you could use [`async.race`](http://caolan.github.io/async/docs.html#race) instead, then you don't have to do the manual success tracking

Comment (dryajov) [reply src/transport.js:57]: Won't `async.race` fail on the first failed attempt? I think we'd want all the addresses to be dialed and be able to inspect those that worked, and not stop on first fail - a la bluebird inspect way, `async.parallel` seems to do just that? Also, not sure if dialing in parallel is the best course of action either, unless we want to open connections on all available transports? At the very least, it should try ALL the ~~transports~~ addresses until one works, the fact that it didn't is a bug.

Comment (jackkleeman) [reply src/transport.js:57]: Currently i think it just takes the first return, which is generally either a success or a timeout, which works well for me. I think dialling in parallel is needed because websocket timeout is around ten seconds, so we really want to try every address at once and then move forward with whatever sticks.

Comment (dryajov) [reply src/transport.js:57]: Mmm... that makes sense. 

Does it makes sense to keep all the available connections around? Would it makes sense to switch between connection depending on latency (probably overkill)?

In fairness tho, the first one to come back is probably the fastest one already 🍭

Comment (jackkleeman) [reply src/transport.js:57]: My thinking exactly

Comment (dryajov) [reply src/transport.js:57]: @dignifiedquire @jackkleeman don't think that either `race` or `parallel` would do what we want in this case. With `race` if the first result is a failure, the callback is invoked even if others might succeed, with `parallel` the callback would be called only after all of the calls are completed, I think we need [async.some](https://caolan.github.io/async/docs.html#some) here 🎢

Comment (dignifiedquire) [reply src/transport.js:57]: I think the proper way to do it is probably this:

- use async.queue with a concurrency of e.g 4
- push all addrs as task to the queue
- start the queue
- in the worker function check if success and if so call `queue.kill` and the callback
- if after finishing the queue no success happened call back in with an error

Comment (dryajov) [reply src/transport.js:57]: @dignifiedquire that should work. I like that we can control the concurrency and not have 20 (figurative and literal 😄 ) dials happen at once.

Comment (jackkleeman) [reply src/transport.js:57]: Cool, I'll leave it up to you to make this change as I'm not well versed in the async functions. I'm glad I could bring the issue to the table :)

Comment (dryajov) [reply src/transport.js:57]: @jackkleeman @dignifiedquire I'm deep into libp2p for the circuit-relay stuff right not, so I can squeeze this in, how quickly do you need this @jackkleeman? Is this something thats blocking you?

Comment (jackkleeman) [reply src/transport.js:57]: It's not blocking me at all as I already maintain a fork of this package for my project (OpenBazaar in the browser), so i will use my simple async.each solution in my fork, and leave it up to you to make a better fix whenever you have time. I would argue it's worth using async.each here in the meantime, as the current implementation is fairly unusable unless you manually filter addresses to be valid, but that's totally up to you guys.

Comment (jackkleeman) [reply src/transport.js:57]: Oh and i don't want you to slow down on the relay as it will be game changing for my project!!

Comment (dryajov) [reply src/transport.js:57]: @jackkleeman - yep, that was a great catch overall, I've patched my local fork with the simpler sequential `next(shift)` so far. This might have been the culprit of a bunch of weird issues I was running into at some point with the the WS vs TCP order of dialing https://github.com/libp2p/js-libp2p-circuit/pull/1#issuecomment-286959112 🥇

Comment (dryajov) [reply src/transport.js:57]: @diasdavid @dignifiedquire +1 to merging this as is for now. 

I've created an issue to track the enhancement proposed by @dignifiedquire

File: src/transport.js
Comment (daviddias) [inline src/transport.js:66]: This is leaving open sockets behind. We need to check: `If I have successfully opened a socket before, then I need to close this one.`

**Review by daviddias — CHANGES_REQUESTED** [review]
Thank you for detecting this and PR'ing a fix @jackkleeman. Just a small comment that needs to be addressed, otherwise LGTM.

File: src/transport.js
Comment (daviddias) [reply src/transport.js:57]: See: https://github.com/libp2p/js-libp2p-swarm/pull/193#discussion_r108138054

Comment (dignifiedquire) [reply src/transport.js:57]: I am going to implement the new dialing today as I need to go into that code anyway

File: src/transport.js
Comment (jackkleeman) [reply src/transport.js:66]: If @dignifiedquire is replacing this PR today should I ignore this?

Comment (dignifiedquire) [reply src/transport.js:66]: Feel free to ignore this yes, I will ping you once the new implementation is up

Comment (jackkleeman) [reply src/transport.js:66]: Thanks! As it turns out, my each solution is failing in a few cases already (when one address fails immeadiately instead of slowly timing out)

**Comment by jackkleeman** [comment]
Outdone by #195

**Pull request closed without merging**
